src/tactic/sort.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2022 Arthur Paulino, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Damiano Testa
+-/
+import data.polynomial.degree.definitions
+
+/-!  # A tactic for sorting sums  -/
+
+namespace tactic
+
+/--  Given an expression `ei` and a list `l : list expr`, returns a `tactic expr` that is
+`ei + sum_of_list l`. -/
+meta def build_sum : expr → list expr → tactic expr
+| ei (e :: es) := do
+  e' ← build_sum ei es,
+  mk_app `has_add.add [e', e]
+| ei [] := pure ei
+
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def get_summands : expr → list expr
+| `(%%a + %%b) := get_summands a ++ get_summands b
+| a            := [a]
+
+section with_weights
+variables {N : Type*} [has_lt N] [decidable_rel (has_lt.lt : N → N → Prop)]
+
+/--  Given an expression `a` and a weight function `wt : expr → N` to a Type `N` with `<`,
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
+order `<`. -/
+meta def sort_summands_with_weight : expr → (expr → N) → list expr
+| a wt := (get_summands a).qsort (λ e f, wt e < wt f)
+
+/--  Let `wt : expr → N` be a "weight function": any function from `expr` to a Type `N` with a
+decidable relation `<`.
+
+Given an expression `e` in an additive commutative semigroup, `sorted_sum_with_weight wt e`
+returns an ordered sum of its terms, where the order is determined by applying `wt` to the summands
+appearing in `e`. -/
+meta def sorted_sum_with_weight (wt : expr → N) (e : expr) : tactic unit :=
+match sort_summands_with_weight e wt with
+| ei::es := do
+  el' ← build_sum ei es.reverse,
+  e_eq ← mk_app `eq [e, el'],
+  n ← get_unused_name,
+  assert n e_eq,
+  -- should we return a meaningful error if this does not manage to prove `e_eq`?
+  reflexivity <|> `[{ simp only [add_comm, add_assoc, add_left_comm], done, }],
```

Comment:
I think you should be able to use the `ac_change` tactic here

Reply:
Ah, I guess I mean `ac_refl`, having looked at the implementation of `ac_change`

Reply:
`ac_refl` loops forever :/

Reply:
yeah, this tactic will be unusable if it uses `ac_refl`

Reply:
Ok, I had left a comment about timings in the example, before knowing about this "feature" of `ac_refl`!